services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java

```diff
@@ -3156,11 +3155,16 @@ private void deleteOneStoreVersion(String clusterName, String storeName, int ver
           cleanUpViewResources(new Properties(), store, deletedVersion.get().getNumber());
         }
         if (store.isDaVinciPushStatusStoreEnabled()) {
-          getPushStatusStoreWriter().deletePushStatus(
-              storeName,
-              deletedVersion.get().getNumber(),
-              Optional.empty(),
-              deletedVersion.get().getPartitionCount());
+          PubSubTopic pushStatusStoreRT = pubSubTopicRepository
+              .getTopic(Version.composeRealTimeTopic(VeniceSystemStoreUtils.getDaVinciPushStatusStoreName(storeName)));
+          if (getTopicManager().containsTopicAndAllPartitionsAreOnline(pushStatusStoreRT)) {
+            getPushStatusStoreWriter().deletePushStatus(storeName, deletedVersion.get().getNumber(), Optional.empty(),
+                deletedVersion.get().getPartitionCount());
```

Comment:
Hmm... even though we check here, there is still a race condition between the time of check and the time of use... Also, the check might be expensive (maybe not a concern for this particular code path, but it is a concern in general, given our Kafka MD load tends to be high already...).

So I wonder if we should make the delete go async in all cases... whether it blocks or not should not block a ZK thread.

Please LMK what you think.

Reply:
Yeah current commit aims to reduce the possibility of producing to a nonexistent topic but does not eliminate it. The chance of topic deletion between topic existence check and writing message to the topic should be very low...I can change the topic existence check to topic truncation check to even lower the possibility.

Regarding "make the delete go async", do you mean that making the "writing delete message to system store RT" part async?

Reply:
Yeah, like... don't even check the topic existence, but execute the delete in some other thread pool, with a max execution time. Do we need to block on this delete before moving forward to the next step of this function? If yes, we can block in-line (but with a timeout), if not, let it go fully async. Or a third option is to block only at the end of the function... IDK what is the right semantic, but in all cases, the point is: let's never allow the code to block forever.

Or anyway, that's just what I'm thinking off the top of my head. Please LMK if you have a concern with this approach.

Reply:
Thanks for the elaboration. The next step of the function is to check and delete user store RT, which should not be blocked by producing system store delete message. IMHO, we still need a timeout in this case. Without timeout, when producing to nonexistent topics, threads will block forever and finally there will be no thread available in the thread pool to produce system store delete message for other stores/versions.

Reply:
I tried a few ways to go async for dvc push status deletion and here are the results:
1. One executor service per class, fire and forget: if a topic does not exist, the thread will wait forever, other stores/versions cannot delete dvc push status.
2. Two executor services per class, one for execution and the other for monitor, fire and forget, the monitor thread waits for 30s and cancel task after timeout: if task1 and task2 are submitted at the same time, and task1 does not finish in 30s, the monitor thread will cancel both task1 and task2, even though task2 is not run at all.
3. One executor service per class, fire, wait, and cancel task after timeout: same problem with 2, task2 could be cancelled without running.
4. One executor service per method call, fire, wait, and cancel task after timeout: stores won't affect each other, thread will not wait forever, executor service will shutdown after timeout or getting the result.

I decided to go with option 4 and updated the PR. When you have time, please take a look and let me know if you have any concern.